### PR TITLE
Ensure `rel=canonical` exists for drafts

### LINKED ISSUE #8162: Add `rel=canonical` link for drafts
## Feature Description

When previewing drafts, users will see `Missing URL for attribute 'href' in tag 'link rel=canonical'.` warnings in the preview mode.

This is because WordPress's [`rel_canonical()`](https://developer.wordpress.org/reference/functions/rel_canonical/)/[`wp_get_canonical_url`](https://developer.wordpress.org/reference/functions/wp_get_canonical_url/) function does not print a canonical tag if a post is not published.

This is confusing.

Example: https://wordpress.org/support/topic/coding-error-with-my-web-story/

## Additional Context

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Implementation Brief

We should update `Canonical_Sanitizer` to use `get_permalink()` instead of `wp_get_canonical_url`.

### PR DESCRIPTION
**Pull request by swissspidy**
## Context

When previewing draft stories, users will see warnings about a missing `rel=canonical` link.

## Summary

This change ensures a `rel=canonical` link is always present, even for drafts.

## Relevant Technical Choices

Simplifies `Canonical_Sanitizer` a bit, adds fallback to `get_permalink()`.

Improves tests.

### QA

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1. Preview a draft story
2. Got to the debug tab in the preview mode
3. Verify that there's no warning about `rel=canonical`

### UAT

- [x] UAT should use the same steps as above.

This PR can be tested by following these steps:

1.

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #8162

### CODE REVIEW

File: includes/AMP/Canonical_Sanitizer.php [hunk 1]
```diff
@@ -54,28 +54,43 @@ class Canonical_Sanitizer extends AMP_Base_Sanitizer {
...
-				$links[ $link->getAttribute( Attribute::REL ) ][] = $link;
+			if ( $link->getAttribute( Attribute::REL ) === Attribute::REL_CANONICAL ) {
+				if ( ! empty( $link->getAttribute( Attribute::HREF ) ) ) {
+					return;
+				}
+
+				$link->setAttribute( Attribute::HREF, (string) $canonical_url );
+
+				break;
 			}
```
Comment (spacedmonkey) [inline includes/AMP/Canonical_Sanitizer.php]: Why does this need to be a foreach loop?

File: includes/AMP/Canonical_Sanitizer.php [hunk 2]
```diff
@@ -54,28 +54,43 @@ class Canonical_Sanitizer extends AMP_Base_Sanitizer {
...
 		 */
 		foreach ( $link_elements as $link ) {
-			if ( $link->hasAttribute( Attribute::REL ) ) {
-				$links[ $link->getAttribute( Attribute::REL ) ][] = $link;
+			if ( $link->getAttribute( Attribute::REL ) === Attribute::REL_CANONICAL ) {
+				if ( ! empty( $link->getAttribute( Attribute::HREF ) ) ) {
+					return;
+				}
+
+				$link->setAttribute( Attribute::HREF, (string) $canonical_url );
```
Comment (spacedmonkey) [inline includes/AMP/Canonical_Sanitizer.php]: Can we check if it is empty before setting it? What if it is false?

File: includes/AMP/Canonical_Sanitizer.php [hunk 3]
```diff
@@ -54,28 +54,43 @@ class Canonical_Sanitizer extends AMP_Base_Sanitizer {
...
+		 */
+		$rel_canonical = null;
 		$link_elements = $this->dom->head->getElementsByTagName( Tag::LINK );
 
+		// This fallback to get_permalink() ensures that there's a canonical link
+		// even when previewing drafts.
+		$canonical_url = wp_get_canonical_url();
+		if ( ! $canonical_url ) {
+			$canonical_url = get_permalink();
+		}
```
Comment (spacedmonkey) [inline includes/AMP/Canonical_Sanitizer.php:65]: If are you viewing as a preview, should this value be `get_preview_post_link`?

Comment (swissspidy) [reply includes/AMP/Canonical_Sanitizer.php:65]: If you just visit a draft directly, e.g. `example.com/?p=12345`, then this wouldn't apply. `get_preview_post_link()` is just `get_permalink()` + `?preview=true` anyway.

So `get_permalink()` is fine IMO.

File: includes/AMP/Canonical_Sanitizer.php (hunk 1 above)
Comment (swissspidy) [reply includes/AMP/Canonical_Sanitizer.php]: It doesn't _need_ to be, but it was a loop already before and seems readable to me.

What would be an alternative? An XPath query?

File: includes/AMP/Canonical_Sanitizer.php (hunk 2 above)
Comment (swissspidy) [reply includes/AMP/Canonical_Sanitizer.php]: According to the docs, `get_permalink()` only returns `false` if the post does not exist, which at this point should never be the case.

Worst case the `href` attribute was empty before and now is still empty, which would be the same situation as before this PR, on `main`.

File: includes/AMP/Canonical_Sanitizer.php (hunk 1 above)
Comment (westonruter) [reply includes/AMP/Canonical_Sanitizer.php]: Right, I think that might be slightly better:

```php
// Ensure link[rel=canonical] exists.
$rel_canonical = $this->dom->xpath->query( './link[rel=canonical]', $this->dom->head )->item( 0 );
if ( ! $rel_canonical instanceof DOMElement ) {
	$rel_canonical = $this->dom->createElement( Tag::LINK );
	$rel_canonical->setAttribute( Attribute::REL, Attribute::REL_CANONICAL );
	$this->dom->head->appendChild( $rel_canonical );
}

// Ensure link[rel=canonical] has a non-empty href attribute.
if ( empty( $rel_canonical->getAttribute( Attribute::HREF ) ) ) {
	$rel_canonical->setAttribute( Attribute::HREF, (string) $canonical_url );
}
```

**Review by spacedmonkey — APPROVED** [review]

**Pull request merged**